linkerd/http/retry/src/compat.rs

```diff
@@ -0,0 +1,115 @@
+//! Compatibility utilities for upgrading to http-body 1.0.
+
+use http_body::Body;
+
+pub(crate) use self::frame::Frame;
+
+mod frame;
+
+#[derive(Debug)]
+pub(crate) struct ForwardCompatibleBody<B> {
+    inner: B,
+    data_finished: bool,
+    trailers_finished: bool,
+}
+
+// === impl ForwardCompatibleBody ===
+
+impl<B: Body> ForwardCompatibleBody<B> {
+    pub(crate) fn new(body: B) -> Self {
+        if body.is_end_stream() {
+            Self {
+                inner: body,
+                data_finished: true,
+                trailers_finished: true,
+            }
+        } else {
+            Self {
+                inner: body,
+                data_finished: false,
+                trailers_finished: false,
+            }
+        }
+    }
+
+    pub(crate) fn into_inner(self) -> B {
+        self.inner
+    }
+
+    /// Returns a future that resolves to the next frame.
+    pub(crate) fn frame(&mut self) -> combinators::Frame<'_, B> {
+        combinators::Frame(self)
+    }
```

Comment:
this gives us a method that lines up with https://docs.rs/http-body-util/0.1.2/src/http_body_util/lib.rs.html#34-39

Reply:
...so we can use this to poll for frames in a forward-compatible way, and then call `into_inner()` to get the `B` back out when we're finished peeking.